jax/core.py

```diff
@@ -398,42 +399,47 @@ def __repr__(self):
       map('  {}\n'.format, self.downward))
 
 
-trace_stack = TraceStack()
-
 class Sublevel(int): pass
-substack = [Sublevel(0)]
+
+_state = threading.local()
```

Comment:
Maybe a new class (and a singleton instance of it) that subclasses `threading.local`, instead of the raw `threading.local` instance.